<<<<<<< HEAD
# EpiGenome-Scale Metabolic Modeling
Still in beta. 
=======
# epiGenome-Scale Metabolic Modeling
Cancer cells have epigenetic and metabolic alterations that support uncontrolled cellular proliferation and growth. Because histone posttranslational modifications (PTMs) are sensitive to altered metabolite concentrations, and specific histone PTMs regulate the expression of metabolic genes, disrupting these altered processes can lead to new cancer therapeutics. 

Experimentally determining important targets in the epigenomic-metabolic network is challenging, time-consuming, and expensive. Thus, new computational algorithms that can uncover and predict the best potential drug targets first are needed to develop new cancer drugs in a time- and cost-efficient manner. 

Constraint-based metabolic models have been used to shed insight into cancer metabolism and predict new drug therapies. However, to date, we are the only group that have developed an epiGenome-Scale Metabolic Model (eGEMMs) of histone acetylation to quantify HDAC inhibitor sensitivity and variability in cancer cells. Because the metabolic and epigenomic network are largely intertwined, we hypothesize that constraint-based modeling can be used to shed insight into the effect histone PTMs have on cancer processes, and use these model predictions to identify potential therapies that target specific PTMs.

The `egem` software package was developed in MATLAB and Python to create new eGEMMs, compute cancer cell metabolic activity using transcriptomic data as the input, and quantify cancer drug sensitivity and variability given cancer transcriptomic data.
>>>>>>> ba4bd1288de2ea80af8a10e9916499807aa3e7a1
